### DISCUSSION TITLE
Split out factory and pool

### DISCUSSION THREAD

**Discussion by damip** [massalabs/massa#2895] [Ideas]
# Split out factory and pool

## General idea

Remove factories and pool from the core, and have core only handle consensus.
Have the removed modules be a separate binary that talks to the node API and handles its own wallet.

Advantages:
* easy to do since the ledger unification refactoring
* better separation of concerns (core now only does consensus)
* more modularity
* the separate binary can be written in any language
* ability to stake from a different computer than the one running a node (to isolate private keys)

![general drawio](https://user-images.githubusercontent.com/10211103/185970105-b21937d1-008b-4768-849c-83c18e2c33c9.svg)

## Changes needed

* remove massa-pool
* remove massa-factory
* add the following endpoints to the API:
  * send endorsement
  * send block
  * get selector draws
  * get previously executed operations
  * get best parents
  * subscribe for streaming of:
    * new endorsements
    * new operations

**Comment by gregLibert** [comment]
This feature will put the block creator back at the center of the stacking logic.
He will be able to:
- choose his own logic for filling a block,
- accept to be paid in the way he wants (native token, other tokens, subscriptions, barter, donations...).

Concerning the protection of keys, and therefore the signature in a secure environment, this point needs to be clarified, especially about performances: are we talking about an application that must run on the same machine, the same subnet?

It is certain that if you can run it on your phone or if you can connect an HSM to it, it changes the game.

Finally, the gain in terms of modularity and attack surface is always good to take :-).

**Reply by damip** [reply]
For keys, the idea is that you get a separate lightweight binary that is the only one managing private keys.
It just talks to a node through its normal API.
You might be able to run the factory + pool directly on an ARM-enabled HSM enclave.

**Comment by damip** [comment]
@jcbaillie  FYI

**Comment by damip** [comment]
WARNING: important caveat: https://github.com/massalabs/massa/pull/3086#discussion_r984636919

We should keep alive a ref to the ops being announced/propagated to avoid dropping them mid-process
